File: backend/app/schemas/geo.py
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


class GeoAnomalyPoint(BaseModel):
    """A geolocated anomaly event projected onto city coordinates."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    lat: float
    lng: float
    severity: str  # CRITICAL, WARNING, INFO
    category: str  # POWER, TRAFFIC, WATER, INTERNET, PUBLIC_INFRASTRUCTURE
    score: float  # Anomaly score [0.0 – 1.0]
    metric_name: str
    description: str
    timestamp: str
    district: str  # Named city zone
    acknowledged: bool = False


class GeoRegionPolygon(BaseModel):
    """GeoJSON-like polygon boundary for a city monitoring zone."""

    region_id: str
    name: str
    coordinates: List[List[float]]  # [[lat, lng], ...] closed polygon


class GeoRegionStatus(BaseModel):
    """Aggregated infrastructure health status for a named city zone."""

    model_config = ConfigDict(from_attributes=True)

    region_id: str
    name: str
    centroid_lat: float
    centroid_lng: float
    health_score: float  # [5.0 – 100.0]
    risk_level: str  # NOMINAL, LOW, MEDIUM, HIGH, CRITICAL
    anomaly_count: int
    critical_count: int
    dominant_category: Optional[str]
    polygon: List[List[float]]  # [[lat, lng], ...]


class GeoHeatmapPoint(BaseModel):
    """A single weighted point for heatmap rendering [lat, lng, intensity]."""

    lat: float
    lng: float
    intensity: float  # [0.0 – 1.0] normalized weight


class GeoMapResponse(BaseModel):
    """Complete geospatial map payload returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    anomaly_points: List[GeoAnomalyPoint]
    regions: List[GeoRegionStatus]
    heatmap_points: List[GeoHeatmapPoint]
    total_anomalies: int
    critical_count: int
    most_affected_region: Optional[str]
    last_updated: str

File: frontend/src/api/geo.ts
import { getWithRetry } from "./client";

// ==============================================================================
// Response Types (mirroring backend Pydantic schemas)
// ==============================================================================

export interface GeoAnomalyPoint {
  id: string;
  lat: number;
  lng: number;
  severity: "CRITICAL" | "WARNING" | "INFO";
  category: "POWER" | "TRAFFIC" | "WATER" | "INTERNET" | "PUBLIC_INFRASTRUCTURE";
  score: number;
  metric_name: string;
  description: string;
  timestamp: string;
  district: string;
  acknowledged: boolean;
}

export interface GeoRegionStatus {
  region_id: string;
  name: string;
  centroid_lat: number;
  centroid_lng: number;
  health_score: number;
  risk_level: "NOMINAL" | "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
  anomaly_count: number;
  critical_count: number;
  dominant_category: string | null;
  polygon: number[][];
}

export interface GeoHeatmapPoint {
  lat: number;
  lng: number;
  intensity: number;
}

export interface GeoMapResponse {
  anomaly_points: GeoAnomalyPoint[];
  regions: GeoRegionStatus[];
  heatmap_points: GeoHeatmapPoint[];
  total_anomalies: number;
  critical_count: number;
  most_affected_region: string | null;
  last_updated: string;
}

export interface GeoAnomalyQueryParams {
  severity?: string;
  category?: string;
  district?: string;
  limit?: number;
}

// ==============================================================================
// API Functions
// ==============================================================================

/**
 * Fetch the complete geospatial map payload.
 */
export async function fetchGeoMap(): Promise<GeoMapResponse> {
  return getWithRetry<GeoMapResponse>("/api/v1/geo/map");
}

/**
 * Fetch per-region infrastructure health status with polygon boundaries.
 */
export async function fetchGeoRegions(): Promise<GeoRegionStatus[]> {
  return getWithRetry<GeoRegionStatus[]>("/api/v1/geo/regions");
}

/**
 * Fetch filtered geolocated anomaly points.
 */
export async function fetchGeoAnomalyPoints(
  params?: GeoAnomalyQueryParams
): Promise<GeoAnomalyPoint[]> {
  return getWithRetry<GeoAnomalyPoint[]>("/api/v1/geo/anomaly-points", { params });
}

/**
 * Fetch heatmap intensity grid points.
 */
export async function fetchGeoHeatmap(): Promise<GeoHeatmapPoint[]> {
  return getWithRetry<GeoHeatmapPoint[]>("/api/v1/geo/heatmap");
}
